core/src/main/java/org/infinispan/interceptors/impl/PrefetchInterceptor.java

```diff
@@ -194,50 +196,44 @@ protected Object handleReadCommand(InvocationContext ctx, DataCommand command) t
+         return asyncInvokeNext(ctx, command, retrieveRemoteValues(ctx, transferedKeys, command.getTopologyId()));
       } else {
          return invokeNext(ctx, command);
       }
    }
 
-   private InvocationStage lookupLocalAndRetrieveRemote(InvocationContext ctx, Object key, TopologyAffectedCommand command, int segment) {
-      if (trace) {
-         log.tracef("Locally prefetching entry for key %s", key);
-      }
-      GetCacheEntryCommand getCacheEntryCommand = commandsFactory.buildGetCacheEntryCommand(key, segment, EnumUtil.bitSetOf(Flag.CACHE_MODE_LOCAL));
-      getCacheEntryCommand.setTopologyId(command.getTopologyId());
-      return makeStage(invokeNextThenApply(ctx, getCacheEntryCommand, handleLocallyLookedUpEntry));
-   }
-
-   private Object handleLocallyLookedUpEntry(InvocationContext ctx1, VisitableCommand command1, Object rv) {
-      GetCacheEntryCommand cmd = (GetCacheEntryCommand) command1;
-      if (trace) {
-         log.tracef("Locally prefetched entry %s", rv);
+   private InvocationStage lookupLocalAndRetrieveRemote(InvocationContext ctx, Object key, TopologyAffectedCommand cmd, int segment) {
+      // We need to lookup the dataContainer directly as GetCacheEntryCommand won't return entry with null value
+      InternalCacheEntry entry = dataContainer.get(key);
+            if (trace) {
+         log.tracef("Locally prefetched entry %s", entry);
       }
-      CacheEntry entry = (CacheEntry) rv;
       Metadata metadata = entry != null ? entry.getMetadata() : null;
       CompletionStage<InternalCacheValue> future;
-      int segment = cmd.getSegment();
       if (metadata != null && metadata.version() != null && svm.isVersionActual(segment, metadata.version())) {
          return null;
       } else if ((metadata instanceof RemoteMetadata) &&
             svm.getSegmentState(segment) == ScatteredVersionManager.SegmentState.VALUE_TRANSFER) {
          // The RemoteMetadata is valid only during value transfer - in blocked state there shouldn't be any such
          // entry and during key transfer we could see metadata pointing to a node with outdated information.
          Address backup = ((RemoteMetadata) metadata).getAddress();
-         future = retrieveRemoteValue(Collections.singleton(backup), cmd.getKey(), segment, cmd);
+         future = retrieveRemoteValue(Collections.singleton(backup), key, segment, cmd);
       } else {
-         future = retrieveRemoteValue(null, cmd.getKey(), segment, cmd);
+         future = retrieveRemoteValue(null, key, segment, cmd);
       }
       return asyncValue(future.thenAccept(maxValue -> {
          if (maxValue == null) {
             return;
          }
+         // The put below could fail updating the context if the data container got the updated value while we were
+         // prefetching that. Also, we need to call RepeatableReadEntry.updatePreviousValue() to get return value
```

Comment:
There's no call to `updatePreviousValue()` in `PrefetchInterceptor`, so I'm not sure how to read this

Reply:
`wrapExternalEntry` calls `updatePreviousValue()` for you. The thing is that we must add the prefetched value to context as the state-transfer-put would not do that if the version is not higher than the one in DC (which happens when we've received the entry in the meantime).